frontend/src/lib/components/TrainTracks.svelte

```diff
@@ -0,0 +1,108 @@
+        "#4e79a7",
+        "#f28e2c",
+        "#e15759",
+        "#76b7b2",
+        "#59a14f",
+        "#edc949",
+        "#af7aa1",
+        "#ff9da7",
+        "#9c755f",
+        "#bab0ab",
+    ];
+    $: colorLength = colors.length;
+    function color(colIdx: number) {
+        return colors[colIdx % colorLength];
+    }
+
+    function calculateCumulativeHeights(simpleHeights: number[]) {
+        let firstRow = (simpleHeights && simpleHeights.length > 0) ? (simpleHeights[0] / 2) : firstRowOffset;
+        let total = firstRow;
+        let result = [firstRow];
+        for (const height of simpleHeights) {
+            total += height;
+            result.push(total);
+        }
+        return result;
+    }
+    $: cumulativeHeights = calculateCumulativeHeights(rowHeights);
+    
+    function rowHeight(rowIdx: number) {
+        return cumulativeHeights[rowIdx] ? cumulativeHeights[rowIdx] : rowHeightDefault * rowIdx + firstRowOffset;
+    }
+
+    function colWidth(colIdx: number) {
+        return colWidthDefault * colIdx + firstColOffset;
+    }
+
+    function bezier(from: Circle, to: Circle) {
+        /* 
+        - If parent was in same column, M (child X,Y) and then V (parent Y)
+        - If no parent, vertical line to bottom of graph (same as above but V (bottom-of-graph Y) instead of parent Y)
+        - If parent was in different column, Bezier curve as follows:
+            - Calculate halfway-point between parent and child. Call it Hx, Hy. Cx, Cy is child, and Px, Py is parent.
+            - M child X,Y
+            - C Cx,Cy to Cx,Hy to Hx,Hy
+            - C Hx,Hy to Px,Hy to Px,Py
+            - Note that parents are *below* children in this graph, so Hy is below Cy and above Py
+            - If I understand that correctly, that could become:
+            - M Cx, Cy
+            - S Cx,Hy Hx,Hy
+            - S Px,Hy Px,Py
+        */
+        if (from && to) {  // TODO: Fix this hack once I figure out why the "to" is sometimes undefined
+        let { row: fromRow, col: fromCol } = from;
+        let { row: toRow, col: toCol } = to;
+        let strokeColor = color(fromCol);  // Consider making it toCol instead
+        let fromX = colWidth(fromCol);
+        let toX = colWidth(toCol);
+        let fromY = rowHeight(fromRow);
+        let toY = rowHeight(toRow);
+        if (fromX == toX) {
+            return { color: strokeColor, d: `M${fromX} ${fromY}V${toY}` };
+        } else {
+            let halfX = (fromX + toX) / 2;
+            let halfY = (fromY + toY) / 2;
+            return { color: strokeColor, d: `M${fromX} ${fromY}S${fromX} ${halfY},${halfX} ${halfY}S${toX} ${halfY},${toX} ${toY}` };
+        }
+    } else return { color: '#000000', d: ''}
+    }
+
+    let curves: SVGStroke[] = [];
+    let svgCircles: SVGDot[] = [];
+
+    $: {
+        cumulativeHeights;  // Lets Svelte know about the hidden dependency on this array in bezier()
```

Comment:
I believe that these properties are also hidden dependencies `rowHeightDefault, firstRowOffset, colWidthDefault, firstColOffset, colors`

Reply:
Yeah, I should probably redo this with stores and `derived` stores to make the dependency relationships more obvious.